docs/source/dataframe-optimizer.rst

```diff
@@ -38,13 +38,25 @@ The optimizations entail the following steps (this list is not complete):
   ``df.groupby(...).apply(...)`` after a merge operation will not shuffle the data
   again if the groupby happens on the merge columns.
 
+  .. figure:: images/optimizer/avoiding-shuffles.svg
+    :align: center
+
+    Performing two subsequent Joins/Merges on the same join-column(s) will avoid shuffling the
+    data again. The optimizer identifies that the partitioning of the DataFrame is already as
+    expected and thus simplifies the operation to a single Shuffle and a trivial merge operation.
```

Comment:
How does this work again? Do we only detect this for subsequent merges or all cases where the data is correctly pre-partitioned?

Reply:
All cases where we partitioned at some point in the past, e.g. 

result = df.merge(df2)
result = result.dropna()
result = result.merge(df3)

would work as well.